storage: TestStoreRangeMergeRHSLeaseExpiration failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/fd520679420038f598c97158672fb201d81ecc13

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=911513&tab=buildLog

```
=== RUN   TestStoreRangeMergeRHSLeaseExpiration
I180918 06:58:50.486463 43084 storage/store.go:1559  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180918 06:58:50.494905 43084 gossip/gossip.go:408  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:37147" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I180918 06:58:50.495750 43371 gossip/gossip.go:940  [n1] gossip connectivity
  n1 [sentinel];
I180918 06:58:50.521979 43084 storage/store.go:1559  [s2] [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
W180918 06:58:50.535061 43084 gossip/gossip.go:1516  [n2] no incoming or outgoing connections
I180918 06:58:50.536012 43084 gossip/gossip.go:408  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:46009" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I180918 06:58:50.537267 43218 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:37147
I180918 06:58:50.537700 43354 gossip/gossip.go:940  [n2] gossip connectivity
I180918 06:58:50.548579 43354 gossip/gossip.go:940  [n2] gossip connectivity
  n1 [sentinel];
  n2 -> n1;
I180918 06:58:50.549627 43371 gossip/gossip.go:940  [n1] gossip connectivity
  n1 [sentinel];
  n2 -> n1;
I180918 06:58:50.616152 43084 storage/replica_command.go:298  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I180918 06:58:50.636569 43084 rpc/nodedialer/nodedialer.go:92  [s1,r1/1:{/Min-b}] connection to n2 established
I180918 06:58:50.641361 43084 storage/store_snapshot.go:615  [s1,r1/1:{/Min-b}] sending preemptive snapshot 8ec059d1 at applied index 18
I180918 06:58:50.642356 43084 storage/store_snapshot.go:657  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 17, log entries: 8, rate-limit: 2.0 MiB/sec, 5ms
I180918 06:58:50.644031 42959 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=8ec059d1, encoded size=2675, 1 rocksdb batches, 8 log entries)
I180918 06:58:50.647781 42959 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=2ms commit=1ms]
I180918 06:58:50.650181 43084 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I180918 06:58:50.656570 43084 storage/replica.go:3754  [s1,r1/1:{/Min-b},txn=aee937e7] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180918 06:58:50.775692 43438 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I180918 06:58:50.804595 43084 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot a2a77f86 at applied index 12
I180918 06:58:50.805826 43084 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 40, log entries: 2, rate-limit: 2.0 MiB/sec, 3ms
I180918 06:58:50.806867 43241 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=a2a77f86, encoded size=7354, 1 rocksdb batches, 2 log entries)
I180918 06:58:50.808641 43241 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I180918 06:58:50.810773 43084 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I180918 06:58:50.821680 43084 storage/replica.go:3754  [s1,r2/1:{b-/Max},txn=8fee3f2e] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180918 06:58:50.864617 43388 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,248 epo=1 pro=0.000000123,249 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,7
I180918 06:58:50.925106 43494 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I180918 06:58:50.950612 43084 storage/client_test.go:1243  test clock advanced to: 1.800000125,0
I180918 06:58:51.655080 43510 storage/client_merge_test.go:1190  starting get 0
I180918 06:58:51.656301 43512 storage/client_merge_test.go:1190  starting get 1
I180918 06:58:51.657517 43513 storage/client_merge_test.go:1190  starting get 2
I180918 06:58:51.658786 43514 storage/client_merge_test.go:1190  starting get 3
I180918 06:58:51.660030 43527 storage/client_merge_test.go:1190  starting get 4
I180918 06:58:51.661254 43528 storage/client_merge_test.go:1190  starting get 5
I180918 06:58:51.662523 43529 storage/client_merge_test.go:1190  starting get 6
I180918 06:58:51.664039 43530 storage/client_merge_test.go:1190  starting get 7
I180918 06:58:51.665673 43511 storage/node_liveness.go:756  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I180918 06:58:51.665984 43460 storage/client_merge_test.go:1190  starting get 8
I180918 06:58:51.666843 43496 storage/client_merge_test.go:1190  starting get 9
I180918 06:58:51.674127 43285 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,142 epo=1 pro=1.800000125,143 following repl=(n2,s2):2 seq=2 start=0.000000123,248 epo=1 pro=0.000000123,249
I180918 06:58:51.742695 43382 storage/store.go:2729  [s2,r1/2:{/Min-b},txn=e05709f5] removing replica r2/2
I180918 06:58:51.752045 43293 storage/store.go:2729  [s1,r1/1:{/Min-b}] removing replica r2/1
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.32s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/14cb3b70da49ba026ad851de7022b14c3ea87acd

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=926889&tab=buildLog

```
I180926 07:05:50.949372 43401 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=2ms commit=1ms]
I180926 07:05:50.952328 43081 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I180926 07:05:50.966761 43081 storage/replica.go:3820  [s1,r1/1:{/Min-b},txn=348cdd92] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180926 07:05:51.074292 43585 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I180926 07:05:51.117608 43081 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 5d08b05b at applied index 12
I180926 07:05:51.118554 43081 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 3ms
I180926 07:05:51.120720 43653 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=5d08b05b, encoded size=7484, 1 rocksdb batches, 2 log entries)
I180926 07:05:51.122711 43653 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I180926 07:05:51.125453 43081 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I180926 07:05:51.140082 43081 storage/replica.go:3820  [s1,r2/1:{b-/Max},txn=ec4e57e7] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180926 07:05:51.196383 43539 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,254 epo=1 pro=0.000000123,255 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,7
I180926 07:05:51.251112 43504 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I180926 07:05:51.293710 43081 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I180926 07:05:52.015503 43616 storage/client_merge_test.go:1190  starting get 0
I180926 07:05:52.016603 43684 storage/client_merge_test.go:1190  starting get 1
I180926 07:05:52.017845 43659 storage/client_merge_test.go:1190  starting get 2
I180926 07:05:52.019140 43685 storage/client_merge_test.go:1190  starting get 3
I180926 07:05:52.020390 43686 storage/client_merge_test.go:1190  starting get 4
I180926 07:05:52.021592 43670 storage/client_merge_test.go:1190  starting get 5
I180926 07:05:52.022899 43687 storage/client_merge_test.go:1190  starting get 6
I180926 07:05:52.024173 43671 storage/client_merge_test.go:1190  starting get 7
I180926 07:05:52.025811 43672 storage/client_merge_test.go:1190  starting get 8
I180926 07:05:52.027224 43673 storage/client_merge_test.go:1190  starting get 9
I180926 07:05:52.027624 43617 storage/node_liveness.go:769  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I180926 07:05:52.041374 43456 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,143 epo=1 pro=1.800000125,144 following repl=(n2,s2):2 seq=2 start=0.000000123,254 epo=1 pro=0.000000123,255
I180926 07:05:52.109614 43532 storage/store.go:2739  [s2,r1/2:{/Min-b},txn=7c01c953] removing replica r2/2
I180926 07:05:52.118275 43435 storage/store.go:2739  [s1,r1/1:{/Min-b}] removing replica r2/1
W180926 07:05:52.128681 43408 storage/raft_transport.go:281  unable to accept Raft message from (n2,s2):?: no handler registered for (n1,s1):?
W180926 07:05:52.132244 43407 storage/raft_transport.go:436  no handler found for store 2 in response range_id:0 from_replica:<node_id:1 store_id:1 replica_id:0 > to_replica:<node_id:2 store_id:2 replica_id:0 > union:<error:<message:"store 1 was not found" transaction_restart:NONE origin_node:0 detail:<store_not_found:<store_id:1 > > now:<> > > 
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.37s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/81022d77ecd7a6532fee9e541f97b995497bbf8a

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=931191&tab=buildLog

```
I180928 06:24:13.324889 43290 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=aca8c1c6, encoded size=2675, 1 rocksdb batches, 8 log entries)
I180928 06:24:13.329231 43290 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=1ms]
I180928 06:24:13.331556 43247 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I180928 06:24:13.338335 43247 storage/replica.go:3761  [s1,r1/1:{/Min-b},txn=e59c6afb] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180928 06:24:13.423115 43622 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I180928 06:24:13.488969 43247 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 305342e7 at applied index 12
I180928 06:24:13.489966 43247 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 2ms
I180928 06:24:13.491207 43676 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=305342e7, encoded size=7442, 1 rocksdb batches, 2 log entries)
I180928 06:24:13.492563 43676 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I180928 06:24:13.494821 43247 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I180928 06:24:13.507680 43247 storage/replica.go:3761  [s1,r2/1:{b-/Max},txn=f9ce0b27] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180928 06:24:13.550569 43574 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,250 epo=1 pro=0.000000123,251 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I180928 06:24:13.606365 43295 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I180928 06:24:13.652925 43247 storage/client_test.go:1243  test clock advanced to: 1.800000125,0
I180928 06:24:14.454274 43678 storage/client_merge_test.go:1190  starting get 0
I180928 06:24:14.455482 43659 storage/client_merge_test.go:1190  starting get 1
I180928 06:24:14.456633 43536 storage/client_merge_test.go:1190  starting get 2
I180928 06:24:14.457971 43660 storage/client_merge_test.go:1190  starting get 3
I180928 06:24:14.459162 43661 storage/client_merge_test.go:1190  starting get 4
I180928 06:24:14.460409 43537 storage/client_merge_test.go:1190  starting get 5
I180928 06:24:14.461760 43680 storage/client_merge_test.go:1190  starting get 6
I180928 06:24:14.463009 43681 storage/client_merge_test.go:1190  starting get 7
I180928 06:24:14.464252 43662 storage/client_merge_test.go:1190  starting get 8
I180928 06:24:14.465441 43648 storage/client_merge_test.go:1190  starting get 9
I180928 06:24:14.465986 43679 storage/node_liveness.go:756  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I180928 06:24:14.474292 43445 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,146 epo=1 pro=1.800000125,147 following repl=(n2,s2):2 seq=2 start=0.000000123,250 epo=1 pro=0.000000123,251
I180928 06:24:14.539217 43569 storage/store.go:2735  [s2,r1/2:{/Min-b},txn=8412dddf] removing replica r2/2
I180928 06:24:14.545635 43457 storage/store.go:2735  [s1,r1/1:{/Min-b}] removing replica r2/1
W180928 06:24:14.547369 43541 storage/node_liveness.go:494  [hb] failed node liveness heartbeat: context canceled
I180928 06:24:14.548613 43663 internal/client/txn.go:629  async rollback failed: node unavailable; try another peer
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.43s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/4e80b4a591764fa75c43b3cb29f384e1c9842a49

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=935456&tab=buildLog

```
I181001 08:36:49.377018 43449 storage/store_snapshot.go:657  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 17, log entries: 8, rate-limit: 2.0 MiB/sec, 5ms
I181001 08:36:49.380568 43713 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=24379a89, encoded size=2675, 1 rocksdb batches, 8 log entries)
I181001 08:36:49.384585 43713 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181001 08:36:49.388980 43449 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I181001 08:36:49.398775 43449 storage/replica.go:3820  [s1,r1/1:{/Min-b},txn=6439cd03] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181001 08:36:49.510827 43686 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181001 08:36:49.552330 43449 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot d94953c3 at applied index 12
I181001 08:36:49.553661 43449 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 5ms
I181001 08:36:49.555205 43468 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=d94953c3, encoded size=7484, 1 rocksdb batches, 2 log entries)
I181001 08:36:49.556976 43468 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181001 08:36:49.560070 43449 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I181001 08:36:49.576351 43449 storage/replica.go:3820  [s1,r2/1:{b-/Max},txn=c55f95e4] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181001 08:36:49.624237 43635 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,257 epo=1 pro=0.000000123,258 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I181001 08:36:49.684696 43342 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I181001 08:36:49.728591 43449 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I181001 08:36:50.505038 43471 storage/client_merge_test.go:1190  starting get 0
I181001 08:36:50.506271 43473 storage/client_merge_test.go:1190  starting get 1
I181001 08:36:50.507547 43766 storage/client_merge_test.go:1190  starting get 2
I181001 08:36:50.508825 43474 storage/client_merge_test.go:1190  starting get 3
I181001 08:36:50.510053 43343 storage/client_merge_test.go:1190  starting get 4
I181001 08:36:50.511274 43676 storage/client_merge_test.go:1190  starting get 5
I181001 08:36:50.512504 43482 storage/client_merge_test.go:1190  starting get 6
I181001 08:36:50.513774 43767 storage/client_merge_test.go:1190  starting get 7
I181001 08:36:50.515062 43483 storage/client_merge_test.go:1190  starting get 8
I181001 08:36:50.516329 43736 storage/client_merge_test.go:1190  starting get 9
I181001 08:36:50.516948 43472 storage/node_liveness.go:769  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I181001 08:36:50.523887 43515 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,146 epo=1 pro=1.800000125,147 following repl=(n2,s2):2 seq=2 start=0.000000123,257 epo=1 pro=0.000000123,258
I181001 08:36:50.590381 43636 storage/store.go:2741  [s2,r1/2:{/Min-b},txn=8d8ec77e] removing replica r2/2
I181001 08:36:50.593193 43543 storage/store.go:2741  [s1,r1/1:{/Min-b}] removing replica r2/1
I181001 08:36:50.595944 43678 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] wait-for-merge
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.42s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/4e7bc613178eeea6b34e31699d9d8878961cfada

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=935520&tab=buildLog

```
I181001 08:51:34.965773 43666 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=c1d8c3e9, encoded size=2675, 1 rocksdb batches, 8 log entries)
I181001 08:51:34.970008 43666 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181001 08:51:34.972972 43419 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I181001 08:51:34.982672 43419 storage/replica.go:3761  [s1,r1/1:{/Min-b},txn=a49a06f9] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181001 08:51:35.099567 43621 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181001 08:51:35.137429 43419 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 04e5dcf9 at applied index 12
I181001 08:51:35.138750 43419 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 4ms
I181001 08:51:35.140772 43684 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=04e5dcf9, encoded size=7442, 1 rocksdb batches, 2 log entries)
I181001 08:51:35.143024 43684 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181001 08:51:35.146637 43419 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I181001 08:51:35.163468 43419 storage/replica.go:3761  [s1,r2/1:{b-/Max},txn=88ba1cc1] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181001 08:51:35.214866 43569 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,251 epo=1 pro=0.000000123,252 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I181001 08:51:35.277409 43686 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I181001 08:51:35.317944 43419 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I181001 08:51:36.150583 43688 storage/client_merge_test.go:1190  starting get 0
I181001 08:51:36.151771 43649 storage/client_merge_test.go:1190  starting get 1
I181001 08:51:36.153027 43690 storage/client_merge_test.go:1190  starting get 2
I181001 08:51:36.154290 43678 storage/client_merge_test.go:1190  starting get 3
I181001 08:51:36.155542 43691 storage/client_merge_test.go:1190  starting get 4
I181001 08:51:36.156842 43527 storage/client_merge_test.go:1190  starting get 5
I181001 08:51:36.158022 43528 storage/client_merge_test.go:1190  starting get 6
I181001 08:51:36.159332 43529 storage/client_merge_test.go:1190  starting get 7
I181001 08:51:36.160582 43692 storage/client_merge_test.go:1190  starting get 8
I181001 08:51:36.161915 43693 storage/client_merge_test.go:1190  starting get 9
I181001 08:51:36.163754 43689 storage/node_liveness.go:756  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I181001 08:51:36.172829 43475 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,149 epo=1 pro=1.800000125,150 following repl=(n2,s2):2 seq=2 start=0.000000123,251 epo=1 pro=0.000000123,252
I181001 08:51:36.243763 43567 storage/store.go:2737  [s2,r1/2:{/Min-b},txn=b6026300] removing replica r2/2
I181001 08:51:36.246330 43483 storage/store.go:2737  [s1,r1/1:{/Min-b}] removing replica r2/1
I181001 08:51:36.249035 43439 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] kv.TxnCoordSender: heartbeat loop
W181001 08:51:36.250154 43623 storage/store.go:1664  [s2,r1/2:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.50s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/f78cc21bb4d7bd3a79e44b4608c04684edf28179

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=940597&tab=buildLog

```
I181003 06:28:53.261191 43430 storage/store_snapshot.go:615  [s1,r1/1:{/Min-b}] sending preemptive snapshot d652a8c9 at applied index 18
I181003 06:28:53.262098 43430 storage/store_snapshot.go:657  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 17, log entries: 8, rate-limit: 2.0 MiB/sec, 3ms
I181003 06:28:53.263488 43441 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=d652a8c9, encoded size=2675, 1 rocksdb batches, 8 log entries)
I181003 06:28:53.266545 43441 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I181003 06:28:53.269021 43430 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I181003 06:28:53.276615 43430 storage/replica.go:3820  [s1,r1/1:{/Min-b},txn=aaf749b5] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181003 06:28:53.406211 43642 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181003 06:28:53.435450 43430 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 860a4294 at applied index 12
I181003 06:28:53.436908 43430 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 11ms
I181003 06:28:53.439520 43426 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=860a4294, encoded size=7484, 1 rocksdb batches, 2 log entries)
I181003 06:28:53.441730 43426 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181003 06:28:53.445421 43430 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I181003 06:28:53.464349 43430 storage/replica.go:3820  [s1,r2/1:{b-/Max},txn=7225daa4] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181003 06:28:53.519296 43474 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,255 epo=1 pro=0.000000123,256 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I181003 06:28:53.585337 43715 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I181003 06:28:53.620517 43430 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I181003 06:28:54.359257 43747 storage/client_merge_test.go:1190  starting get 0
I181003 06:28:54.360436 43665 storage/client_merge_test.go:1190  starting get 1
I181003 06:28:54.361641 43748 storage/client_merge_test.go:1190  starting get 2
I181003 06:28:54.362877 43749 storage/client_merge_test.go:1190  starting get 3
I181003 06:28:54.364144 43666 storage/client_merge_test.go:1190  starting get 4
I181003 06:28:54.365425 43716 storage/client_merge_test.go:1190  starting get 5
I181003 06:28:54.366675 43750 storage/client_merge_test.go:1190  starting get 6
I181003 06:28:54.367907 43704 storage/client_merge_test.go:1190  starting get 7
I181003 06:28:54.369171 43763 storage/client_merge_test.go:1190  starting get 8
I181003 06:28:54.370440 43686 storage/client_merge_test.go:1190  starting get 9
I181003 06:28:54.372475 43703 storage/node_liveness.go:769  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I181003 06:28:54.381124 43486 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,146 epo=1 pro=1.800000125,147 following repl=(n2,s2):2 seq=2 start=0.000000123,255 epo=1 pro=0.000000123,256
I181003 06:28:54.457293 43471 storage/store.go:2741  [s2,r1/2:{/Min-b},txn=d480d551] removing replica r2/2
I181003 06:28:54.463434 43493 storage/store.go:2741  [s1,r1/1:{/Min-b}] removing replica r2/1
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.39s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/6f93127c54588e2afcebeaee165ed2496f1d3705

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=943236&tab=buildLog

```
I181004 07:57:19.080121 43479 storage/store_snapshot.go:615  [s1,r1/1:{/Min-b}] sending preemptive snapshot c170a74b at applied index 18
I181004 07:57:19.081290 43479 storage/store_snapshot.go:657  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 17, log entries: 8, rate-limit: 2.0 MiB/sec, 4ms
I181004 07:57:19.082862 43686 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=c170a74b, encoded size=2677, 1 rocksdb batches, 8 log entries)
I181004 07:57:19.086823 43686 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181004 07:57:19.089950 43479 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I181004 07:57:19.100839 43479 storage/replica.go:3763  [s1,r1/1:{/Min-b},txn=f2358b50] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181004 07:57:19.226448 43659 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181004 07:57:19.252572 43479 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 2856b1e7 at applied index 12
I181004 07:57:19.253732 43479 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 4ms
I181004 07:57:19.255721 43732 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=2856b1e7, encoded size=7442, 1 rocksdb batches, 2 log entries)
I181004 07:57:19.257244 43732 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I181004 07:57:19.260747 43479 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I181004 07:57:19.281931 43479 storage/replica.go:3763  [s1,r2/1:{b-/Max},txn=48f1a5b6] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181004 07:57:19.346973 43611 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,249 epo=1 pro=0.000000123,250 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I181004 07:57:19.404851 43707 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I181004 07:57:19.439037 43479 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I181004 07:57:20.265036 43747 storage/client_merge_test.go:1190  starting get 0
I181004 07:57:20.266332 43677 storage/client_merge_test.go:1190  starting get 1
I181004 07:57:20.267677 43709 storage/client_merge_test.go:1190  starting get 2
I181004 07:57:20.269031 43749 storage/client_merge_test.go:1190  starting get 3
I181004 07:57:20.270276 43710 storage/client_merge_test.go:1190  starting get 4
I181004 07:57:20.271593 43694 storage/client_merge_test.go:1190  starting get 5
I181004 07:57:20.272852 43695 storage/client_merge_test.go:1190  starting get 6
I181004 07:57:20.274120 43678 storage/client_merge_test.go:1190  starting get 7
I181004 07:57:20.275424 43679 storage/client_merge_test.go:1190  starting get 8
I181004 07:57:20.276666 43723 storage/client_merge_test.go:1190  starting get 9
I181004 07:57:20.277136 43748 storage/node_liveness.go:756  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I181004 07:57:20.286034 43508 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,150 epo=1 pro=1.800000125,151 following repl=(n2,s2):2 seq=2 start=0.000000123,249 epo=1 pro=0.000000123,250
I181004 07:57:20.356182 43608 storage/store.go:2737  [s2,r1/2:{/Min-b},txn=04621530] removing replica r2/2
I181004 07:57:20.357239 43517 storage/store.go:2737  [s1,r1/1:{/Min-b}] removing replica r2/1
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.47s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/93a54f9be34d00b755e440bd76240a19f04650b5

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=945479&tab=buildLog

```
I181005 06:24:15.667269 43624 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I181005 06:24:15.669742 43379 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I181005 06:24:15.677273 43379 storage/replica.go:3763  [s1,r1/1:{/Min-b},txn=8e05a7df] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181005 06:24:15.796463 43369 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181005 06:24:15.829056 43379 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 81c76a4e at applied index 12
I181005 06:24:15.830409 43379 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 4ms
I181005 06:24:15.832296 43590 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=81c76a4e, encoded size=7442, 1 rocksdb batches, 2 log entries)
I181005 06:24:15.833824 43590 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I181005 06:24:15.836886 43379 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I181005 06:24:15.851957 43379 storage/replica.go:3763  [s1,r2/1:{b-/Max},txn=9499943c] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181005 06:24:15.905283 43519 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,251 epo=1 pro=0.000000123,252 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I181005 06:24:15.960164 43505 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I181005 06:24:15.998485 43379 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I181005 06:24:16.044113 43591 storage/client_merge_test.go:1190  starting get 0
I181005 06:24:16.045210 43471 storage/client_merge_test.go:1190  starting get 1
I181005 06:24:16.046505 43669 storage/client_merge_test.go:1190  starting get 2
I181005 06:24:16.047755 43610 storage/client_merge_test.go:1190  starting get 3
I181005 06:24:16.053812 43670 storage/client_merge_test.go:1190  starting get 4
I181005 06:24:16.055143 43211 storage/client_merge_test.go:1190  starting get 5
I181005 06:24:16.056530 43671 storage/client_merge_test.go:1190  starting get 6
I181005 06:24:16.057744 43472 storage/client_merge_test.go:1190  starting get 7
I181005 06:24:16.059377 43672 storage/client_merge_test.go:1190  starting get 8
I181005 06:24:16.060542 43593 storage/client_merge_test.go:1190  starting get 9
I181005 06:24:16.062860 43592 storage/node_liveness.go:756  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I181005 06:24:16.071590 43390 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,64 epo=1 pro=1.800000125,67 following repl=(n2,s2):2 seq=2 start=0.000000123,251 epo=1 pro=0.000000123,252
I181005 06:24:16.144875 43562 storage/store.go:2737  [s2,r1/2:{/Min-b},txn=772b83e5] removing replica r2/2
I181005 06:24:16.147612 43647 storage/store.go:3815  [s1,r2/1:{b-/Max}] added to replica GC queue (contacted deleted peer)
I181005 06:24:16.149531 43399 storage/store.go:2737  [s1,r1/1:{/Min-b}] removing replica r2/1
I181005 06:24:16.150035 43654 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] storage.replicaGC: processing replica
W181005 06:24:16.169192 43645 storage/raft_transport.go:583  while processing outgoing Raft queue to node 2: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (0.73s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/32346197937265bbc71dd3629b1137df9bf16fb4

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=948671&tab=buildLog

```
I181007 07:49:59.708670 43963 storage/replica_command.go:298  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I181007 07:49:59.732665 43963 storage/store_snapshot.go:615  [s1,r1/1:{/Min-b}] sending preemptive snapshot 9b1ed7f5 at applied index 18
I181007 07:49:59.733924 43963 storage/store_snapshot.go:657  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 17, log entries: 8, rate-limit: 2.0 MiB/sec, 4ms
I181007 07:49:59.735647 43954 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=9b1ed7f5, encoded size=2677, 1 rocksdb batches, 8 log entries)
I181007 07:49:59.738343 43954 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I181007 07:49:59.741421 43963 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I181007 07:49:59.748466 43963 storage/replica.go:3825  [s1,r1/1:{/Min-b},txn=85816e9c] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181007 07:49:59.904086 43963 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot 0a175cbf at applied index 12
I181007 07:49:59.905488 43963 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 6ms
I181007 07:49:59.907144 43982 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=0a175cbf, encoded size=7484, 1 rocksdb batches, 2 log entries)
I181007 07:49:59.909505 43982 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181007 07:49:59.912891 43963 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I181007 07:49:59.927919 43963 storage/replica.go:3825  [s1,r2/1:{b-/Max},txn=a79d388f] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181007 07:49:59.976804 44122 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,279 epo=1 pro=0.000000123,281 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,6
I181007 07:50:00.039354 44260 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I181007 07:50:00.078713 43963 storage/client_test.go:1252  test clock advanced to: 1.800000125,0
I181007 07:50:00.106973 44275 storage/client_merge_test.go:1190  starting get 0
I181007 07:50:00.108249 44291 storage/client_merge_test.go:1190  starting get 1
I181007 07:50:00.109568 44292 storage/client_merge_test.go:1190  starting get 2
I181007 07:50:00.110865 44293 storage/client_merge_test.go:1190  starting get 3
I181007 07:50:00.112141 44229 storage/client_merge_test.go:1190  starting get 4
I181007 07:50:00.113594 44230 storage/client_merge_test.go:1190  starting get 5
I181007 07:50:00.114814 44294 storage/client_merge_test.go:1190  starting get 6
I181007 07:50:00.116097 44277 storage/client_merge_test.go:1190  starting get 7
I181007 07:50:00.117349 44231 storage/client_merge_test.go:1190  starting get 8
I181007 07:50:00.118566 44220 storage/client_merge_test.go:1190  starting get 9
I181007 07:50:00.121475 44276 storage/node_liveness.go:769  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I181007 07:50:00.139634 44008 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,65 epo=1 pro=1.800000125,66 following repl=(n2,s2):2 seq=2 start=0.000000123,279 epo=1 pro=0.000000123,281
I181007 07:50:00.212661 44167 storage/store.go:2741  [s2,r1/2:{/Min-b},txn=151e7cb6] removing replica r2/2
I181007 07:50:00.215503 44020 storage/store.go:2741  [s1,r1/1:{/Min-b}] removing replica r2/1
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (0.72s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

--- a/pkg/roachpb/api.pb.go
+++ b/pkg/roachpb/api.pb.go
@@ -515,7 +515,7 @@ func (*GetRequest) ProtoMessage()               {}
 func (*GetRequest) Descriptor() ([]byte, []int) { return fileDescriptorApi, []int{3} }
 
 // A GetResponse is the return value from the Get() method.
-// If the key doesn't exist, returns nil for Value.Bytes.
+// If the key doesn't exist, Value will be nil.
 type GetResponse struct {
 	ResponseHeader `protobuf:"bytes,1,opt,name=header,embedded=header" json:"header"`
 	Value          *Value `protobuf:"bytes,2,opt,name=value" json:"value,omitempty"`

--- a/pkg/roachpb/api.proto
+++ b/pkg/roachpb/api.proto
@@ -120,7 +120,7 @@ message GetRequest {
 }
 
 // A GetResponse is the return value from the Get() method.
-// If the key doesn't exist, returns nil for Value.Bytes.
+// If the key doesn't exist, Value will be nil.
 message GetResponse {
   ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
   Value value = 2;

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -1972,6 +1972,246 @@ func TestStoreRangeMergeReadoptedLHSFollower(t *testing.T) {
 	mtc.transferLease(ctx, lhsDesc.RangeID, 0, 2)
 }
 
+// TestStoreRangeMergeWatcher verifies that the watcher goroutine for a merge's
+// RHS does not erroneously permit traffic after the merge commits.
+func TestStoreRangeMergeWatcher(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	testutils.RunTrueAndFalse(t, "inject-failures", testMergeWatcher)
+}
+
+func testMergeWatcher(t *testing.T, injectFailures bool) {
+	ctx := context.Background()
+	storeCfg := storage.TestStoreConfig(nil)
+	storeCfg.TestingKnobs.DisableReplicateQueue = true
+	storeCfg.TestingKnobs.DisableReplicaGCQueue = true
+
+	var mergeTxnRetries, pushTxnRetries, meta2GetRetries int64
+	if injectFailures {
+		mergeTxnRetries = 3
+		pushTxnRetries = 3
+		meta2GetRetries = 3
+	}
+
+	// Maybe inject some retryable errors when the merge transaction commits.
+	var mtc *multiTestContext
+	storeCfg.TestingKnobs.TestingRequestFilter = func(ba roachpb.BatchRequest) *roachpb.Error {
+		for _, req := range ba.Requests {
+			if et := req.GetEndTransaction(); et != nil && et.InternalCommitTrigger.GetMergeTrigger() != nil {
+				if atomic.AddInt64(&mergeTxnRetries, -1) >= 0 {
+					return roachpb.NewError(roachpb.NewTransactionRetryError(roachpb.RETRY_SERIALIZABLE))
+				}
+			}
+			if pt := req.GetPushTxn(); pt != nil {
+				if atomic.AddInt64(&pushTxnRetries, -1) >= 0 {
+					return roachpb.NewErrorf("injected failure")
+				}
+			}
+			if g := req.GetGet(); g != nil && ba.ReadConsistency == roachpb.READ_UNCOMMITTED {
+				if atomic.AddInt64(&meta2GetRetries, -1) >= 0 {
+					return roachpb.NewErrorf("injected failure")
+				}
+			}
+		}
+		return nil
+	}
+
+	mtc = &multiTestContext{storeConfig: &storeCfg}
+	mtc.Start(t, 3)
+	defer mtc.Stop()
+	store0, store2 := mtc.Store(0), mtc.Store(2)
+
+	// Make store0 the leaseholder of the LHS and store2 the leaseholder of the
+	// RHS. We'll be forcing store2's LHS to fall behind. This creates an
+	// interesting scenario in which the leaseholder for the RHS has very
+	// out-of-date information about the status of the merge.
+	mtc.replicateRange(roachpb.RangeID(1), 1, 2)
+	lhsDesc, rhsDesc, err := createSplitRanges(ctx, store0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	mtc.transferLease(ctx, rhsDesc.RangeID, 0, 2)
+
+	// Block Raft traffic to the LHS replica on store2, by holding its raftMu, so
+	// that its LHS isn't aware there's a merge in progress.
+	lhsRepl2, err := store2.GetReplica(lhsDesc.RangeID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lhsRepl2.RaftLock()
+
+	args := adminMergeArgs(lhsDesc.StartKey.AsRawKey())
+	_, pErr := client.SendWrapped(ctx, store0.TestSender(), args)
+	if pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Immediately after the merge completes, send a request to the RHS which will
+	// be handled by the leaseholder, on store2. This exercises a tricky scenario.
+	// We've forced store2's LHS replica to fall behind, so it can't subsume
+	// store2's RHS. store2's RHS is watching for the merge to complete, however,
+	// and will notice that the merge has committed before the LHS does.
+	getErr := make(chan error)
+	go func() {
+		_, pErr = client.SendWrappedWith(ctx, store2.TestSender(), roachpb.Header{
+			RangeID: rhsDesc.RangeID,
+		}, getArgs(rhsDesc.StartKey.AsRawKey()))
+		getErr <- pErr.GoError()
+	}()
+
+	// Restore communication with store2. Give it the lease to force all commands
+	// to be applied, including the merge trigger.
+	lhsRepl2.RaftUnlock()
+	mtc.transferLease(ctx, lhsDesc.RangeID, 0, 2)
+
+	// We *must* see a RangeNotFound error from the get request we sent earlier
+	// because we sent it after the merge completed. Anything else is a
+	// consistency error (or a bug in the test).
+	if err := <-getErr; !testutils.IsError(err, "r2 was not found") {
+		t.Fatalf("expected RangeNotFound error from get after merge, but got %v", err)
+	}
+}
+
+// TestStoreRangeMergeSlowWatcher verifies that the watcher goroutine for the
+// RHS of a merge does not erroneously permit traffic after the merge commits,
+// even if the watcher goroutine is so slow in noticing the merge that another
+// merge occurs.
+//
+// This test is a more complicated version of TestStoreRangeMergeWatcher that
+// exercises a rare but important edge case.
+//
+// The test creates three ranges, [a, b), [b, c), and [c, /Max). Hereafter these
+// ranges will be referred to as A, B, and C, respectively. store0 holds the
+// lease on A and C, while store1 holds the lease on B. The test will execute
+// two merges such that first A subsumes B, then AB subsumes C. The idea is to
+// inform store1 that the A <- B merge is in progress so that it locks B down,
+// but then keep it in the dark about the status of the merge for long enough
+// that the AB <- C merge commits.
+//
+// When store1's merge watcher goroutine looks up whether the A <- B merge
+// commit occurred in meta2 with a Get(/Meta2/c) request, it won't find the
+// descriptor for B, which would indicate that the merge aborted, nor the
+// descriptor for AB, which would indicate that the merge committed. Instead it
+// will find no descriptor at all, since the AB <- C merge has committed and the
+// descriptor for the merged range ABC is stored at /Meta2/Max, not /Meta2/c.
+func TestStoreRangeMergeSlowWatcher(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	t.Skip("flawed test: deadlocks if merge transaction retries")
+
+	ctx := context.Background()
+	aKey, bKey, cKey := roachpb.RKey("a"), roachpb.RKey("b"), roachpb.RKey("c")
+	storeCfg := storage.TestStoreConfig(nil)
+	storeCfg.TestingKnobs.DisableReplicateQueue = true
+	var mtc *multiTestContext
+
+	// Force PushTxn requests generated by the watcher goroutine to wait on a
+	// channel. This is how we control when store1's merge watcher goroutine hears
+	// about the status of the A <- B merge.
+	allowPushTxn := make(chan struct{})
+	storeCfg.TestingKnobs.TestingRequestFilter = func(ba roachpb.BatchRequest) *roachpb.Error {
+		for _, req := range ba.Requests {
+			// We can detect PushTxn requests generated by the watcher goroutine
+			// because they use the minimum transaction priority.
+			if pt := req.GetPushTxn(); pt != nil && pt.PusherTxn.Priority == roachpb.MinTxnPriority {
+				<-allowPushTxn
+			}
+		}
+		return nil
+	}
+
+	// Record whether we've seen a request to Get(/Meta2/c) that returned nil.
+	// This verifies that we're actually testing what we claim to.
+	var sawMeta2Req int64
+	meta2CKey := keys.RangeMetaKey(cKey).AsRawKey()
+	storeCfg.TestingKnobs.TestingResponseFilter = func(ba roachpb.BatchRequest, br *roachpb.BatchResponse) *roachpb.Error {
+		for i, req := range ba.Requests {
+			if g := req.GetGet(); g != nil && g.Key.Equal(meta2CKey) && br.Responses[i].GetGet().Value == nil {
+				atomic.StoreInt64(&sawMeta2Req, 1)
+			}
+		}
+		return nil
+	}
+
+	mtc = &multiTestContext{storeConfig: &storeCfg}
+	mtc.Start(t, 3)
+	defer mtc.Stop()
+	store0, store1 := mtc.Store(0), mtc.Store(1)
+
+	// Create and place the ranges as described in the comment on this test.
+	mtc.replicateRange(roachpb.RangeID(1), 1, 2)
+	keys := []roachpb.RKey{aKey, bKey, cKey}
+	for _, key := range keys {
+		splitArgs := adminSplitArgs(key.AsRawKey())
+		if _, pErr := client.SendWrapped(ctx, mtc.distSenders[0], splitArgs); pErr != nil {
+			t.Fatal(pErr)
+		}
+	}
+	bRangeID := store0.LookupReplica(bKey).RangeID
+	mtc.transferLease(ctx, bRangeID, 0, 1)
+
+	// Warm the DistSender cache on each node. We'll be blocking requests to B
+	// during the test, and we don't want requests headed for A or C to get routed
+	// to B while its blocked because of a stale DistSender cache.
+	for _, key := range keys {
+		for _, distSender := range mtc.distSenders {
+			if _, pErr := client.SendWrapped(ctx, distSender, getArgs(key.AsRawKey())); pErr != nil {
+				t.Fatal(pErr)
+			}
+		}
+	}
+
+	// Force the replica of A on store1 to fall behind so that it doesn't apply
+	// any merge triggers. This makes the watcher goroutine responsible for
+	// marking B as destroyed.
+	aRepl1 := store1.LookupReplica(aKey)
+	aRepl1.RaftLock()
+	defer aRepl1.RaftUnlock()
+
+	// Merge A <- B.
+	mergeArgs := adminMergeArgs(aKey.AsRawKey())
+	if _, pErr := client.SendWrapped(ctx, mtc.distSenders[0], mergeArgs); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Immediately after the merge completes, send a request to B.
+	getErr := make(chan error)
+	go func() {
+		_, pErr := client.SendWrappedWith(ctx, store1.TestSender(), roachpb.Header{
+			RangeID: bRangeID,
+		}, getArgs(bKey.AsRawKey()))
+		getErr <- pErr.GoError()
+	}()
+
+	// Merge AB <- C.
+	if _, pErr := client.SendWrapped(ctx, mtc.distSenders[0], mergeArgs); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Synchronously ensure that the intent on meta2CKey has been cleaned up.
+	// The merge committed, but the intent resolution happens asynchronously.
+	_, pErr := client.SendWrapped(ctx, mtc.distSenders[0], getArgs(meta2CKey))
+	if pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// With the meta2CKey intent cleaned up, allow store1's merge watcher
+	// goroutine to proceed.
+	close(allowPushTxn)
+
+	// We *must* see a RangeNotFound error from the get request we sent earlier
+	// because we sent it after the merge completed. Anything else is a
+	// consistency error (or a bug in the test).
+	expErr := fmt.Sprintf("r%d was not found", bRangeID)
+	if err := <-getErr; !testutils.IsError(err, expErr) {
+		t.Fatalf("expected %q error from get after merge, but got %v", expErr, err)
+	}
+
+	if atomic.LoadInt64(&sawMeta2Req) != 1 {
+		t.Fatalf("test did not generate expected meta2 get request/response")
+	}
+}
+
 // unreliableRaftHandler drops all Raft messages that are addressed to the
 // specified rangeID, but lets all other messages through.
 type unreliableRaftHandler struct {

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -228,6 +228,9 @@ const (
 	destroyReasonRemovalPending
 	// The replica has been GCed.
 	destroyReasonRemoved
+	// The replica has been merged into its left-hand neighbor, but its left-hand
+	// neighbor hasn't yet subsumed it.
+	destroyReasonMergePending
 )
 
 type destroyStatus struct {
@@ -2855,8 +2858,9 @@ func (r *Replica) maybeWatchForMerge(ctx context.Context) error {
 	} else if len(intents) > 1 {
 		log.Fatalf(ctx, "MVCCGet returned an impossible number of intents (%d)", len(intents))
 	}
+	intent := intents[0]
 	val, _, err := engine.MVCCGetAsTxn(
-		ctx, r.Engine(), descKey, intents[0].Txn.Timestamp, intents[0].Txn)
+		ctx, r.Engine(), descKey, intent.Txn.Timestamp, intent.Txn)
 	if err != nil {
 		return err
 	} else if val != nil {
@@ -2882,6 +2886,7 @@ func (r *Replica) maybeWatchForMerge(ctx context.Context) error {
 
 	taskCtx := r.AnnotateCtx(context.Background())
 	err = r.store.stopper.RunAsyncTask(taskCtx, "wait-for-merge", func(ctx context.Context) {
+		var pushTxnRes *roachpb.PushTxnResponse
 		for retry := retry.Start(base.DefaultRetryOptions()); retry.Next(); {
 			// Wait for the merge transaction to complete by attempting to push it. We
 			// don't want to accidentally abort the merge transaction, so we use the
@@ -2889,38 +2894,106 @@ func (r *Replica) maybeWatchForMerge(ctx context.Context) error {
 			// roachpb.PUSH_TOUCH, though it might appear more semantically correct,
 			// returns immediately and causes us to spin hot, whereas
 			// roachpb.PUSH_ABORT efficiently blocks until the transaction completes.
-			_, pErr := client.SendWrapped(ctx, r.DB().NonTransactionalSender(), &roachpb.PushTxnRequest{
-				RequestHeader: roachpb.RequestHeader{Key: intents[0].Txn.Key},
+			res, pErr := client.SendWrapped(ctx, r.DB().NonTransactionalSender(), &roachpb.PushTxnRequest{
+				RequestHeader: roachpb.RequestHeader{Key: intent.Txn.Key},
 				PusherTxn: roachpb.Transaction{
 					TxnMeta: enginepb.TxnMeta{Priority: roachpb.MinTxnPriority},
 				},
-				PusheeTxn: intents[0].Txn,
+				PusheeTxn: intent.Txn,
 				Now:       r.Clock().Now(),
 				PushType:  roachpb.PUSH_ABORT,
 			})
 			if pErr != nil {
 				select {
 				case <-r.store.stopper.ShouldQuiesce():
-					// The server is shutting down. The error while fetching the range
-					// descriptor was probably caused by the shutdown, so ignore it.
+					// The server is shutting down. The error while pushing the
+					// transaction was probably caused by the shutdown, so ignore it.
 					return
 				default:
-					log.Warningf(ctx, "error while watching for merge to complete: %s", pErr)
+					log.Warningf(ctx, "error while watching for merge to complete: PushTxn: %s", pErr)
 					// We can't safely unblock traffic until we can prove that the merge
 					// transaction is committed or aborted. Nothing to do but try again.
 					continue
 				}
 			}
-			// Unblock pending requests. If the merge committed, the requests will
-			// notice that the replica has been destroyed and return an appropriate
-			// error. If the merge aborted, the requests will be handled normally.
-			r.mu.Lock()
-			r.mu.mergeComplete = nil
-			close(mergeCompleteCh)
-			r.mu.Unlock()
-			return
+			pushTxnRes = res.(*roachpb.PushTxnResponse)
+			break
 		}
-		log.Fatal(ctx, "unreachable")
+
+		var mergeCommitted bool
+		switch pushTxnRes.PusheeTxn.Status {
+		case roachpb.PENDING:
+			log.Fatalf(ctx, "PushTxn returned while merge transaction %s was still pending",
+				intent.Txn.ID.Short())
+		case roachpb.COMMITTED:
+			// If PushTxn claims that the transaction committed, then the transaction
+			// definitely committed.
+			mergeCommitted = true
+		case roachpb.ABORTED:
+			// If PushTxn claims that the transaction aborted, it's not a guarantee
+			// that the transaction actually aborted. It could also mean that the
+			// transaction completed, resolved its intents, and GC'd its transaction
+			// record before our PushTxn arrived. To figure out what happened, we
+			// need to look in meta2.
+			var getRes *roachpb.GetResponse
+			for retry := retry.Start(base.DefaultRetryOptions()); retry.Next(); {
+				metaKey := keys.RangeMetaKey(desc.EndKey)
+				res, pErr := client.SendWrappedWith(ctx, r.DB().NonTransactionalSender(), roachpb.Header{
+					// Use READ_UNCOMMITTED to avoid trying to resolve intents, since
+					// resolving those intents might involve sending requests to this
+					// range, and that could deadlock. See the comment on
+					// TestStoreRangeMergeConcurrentSplit for details.
+					ReadConsistency: roachpb.READ_UNCOMMITTED,
+				}, &roachpb.GetRequest{
+					RequestHeader: roachpb.RequestHeader{Key: metaKey.AsRawKey()},
+				})
+				if pErr != nil {
+					select {
+					case <-r.store.stopper.ShouldQuiesce():
+						// The server is shutting down. The error while fetching the range
+						// descriptor was probably caused by the shutdown, so ignore it.
+						return
+					default:
+						log.Warningf(ctx, "error while watching for merge to complete: Get %s: %s", metaKey, pErr)
+						// We can't safely unblock traffic until we can prove that the merge
+						// transaction is committed or aborted. Nothing to do but try again.
+						continue
+					}
+				}
+				getRes = res.(*roachpb.GetResponse)
+				break
+			}
+			if getRes.Value == nil {
+				// A range descriptor with our end key is no longer present in meta2, so
+				// the merge must have committed.
+				mergeCommitted = true
+			} else {
+				// A range descriptor with our end key is still present in meta2. The
+				// merge committed iff that range descriptor has a different range ID.
+				var meta2Desc roachpb.RangeDescriptor
+				if err := getRes.Value.GetProto(&meta2Desc); err != nil {
+					log.Fatalf(ctx, "error while watching for merge to complete: "+
+						"unmarshaling meta2 range descriptor: %s", err)
+				}
+				if meta2Desc.RangeID != r.RangeID {
+					mergeCommitted = true
+				}
+			}
+		}
+
+		r.mu.Lock()
+		if mergeCommitted && r.mu.destroyStatus.IsAlive() {
+			// The merge committed but the left-hand replica on this store hasn't
+			// subsumed this replica yet. Mark this replica as destroyed so it
+			// doesn't serve requests when we close the mergeCompleteCh below.
+			r.mu.destroyStatus.Set(roachpb.NewRangeNotFoundError(r.RangeID), destroyReasonMergePending)
+		}
+		// Unblock pending requests. If the merge committed, the requests will
+		// notice that the replica has been destroyed and return an appropriate
+		// error. If the merge aborted, the requests will be handled normally.
+		r.mu.mergeComplete = nil
+		close(mergeCompleteCh)
+		r.mu.Unlock()
 	})
 	if err == stop.ErrUnavailable {
 		// We weren't able to launch a goroutine to watch for the merge's completion

